pallets/did/src/identity_consumer.rs

```diff
@@ -0,0 +1,204 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2022 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use kilt_support::traits::{IdentityConsumer, IdentityCounter, IdentityDecrementer, IdentityIncrementer};
+use sp_runtime::DispatchError;
+
+use crate::{Config, DidConsumers, DidIdentifierOf, Error, Pallet, WeightInfo};
+
+#[derive(Debug, PartialEq)]
+pub struct DidIncrementer<T>(T);
+
+impl<T: Config, Identity> IdentityCounter<u32> for DidIncrementer<(Option<T>, Identity)>
+where
+	// FIXME: remove dependency
+	// on Clone, and change Into
+	// to AsRef, if possible
+	Identity: Into<DidIdentifierOf<T>> + Clone,
+{
+	fn current_value(&self) -> u32 {
+		DidConsumers::<T>::get(&self.0 .1.clone().into())
+	}
+}
+
+impl<T: Config, Identity> IdentityIncrementer<u32> for DidIncrementer<(Option<T>, Identity)>
+where
+	// FIXME: remove dependency
+	// on Clone, and change Into
+	// to AsRef, if possible
+	Identity: Into<DidIdentifierOf<T>> + Clone,
+{
+	fn increment(&mut self) -> frame_support::dispatch::Weight {
+		Pallet::<T>::increment_consumers_unsafe(&self.0 .1.clone().into());
+		T::WeightInfo::increment_consumers()
+	}
+}
+
+#[derive(Debug, PartialEq)]
+pub struct DidDecrementer<T>(T);
+
+impl<T: Config, Identity> IdentityCounter<u32> for DidDecrementer<(Option<T>, Identity)>
+where
+	// FIXME: remove dependency
+	// on Clone, and change Into
+	// to AsRef, if possible
+	Identity: Into<DidIdentifierOf<T>> + Clone,
+{
+	fn current_value(&self) -> u32 {
+		DidConsumers::<T>::get(&self.0 .1.clone().into())
+	}
+}
+
+impl<T: Config, Identity> IdentityDecrementer<u32> for DidDecrementer<(Option<T>, Identity)>
+where
+	// FIXME: remove dependency
+	// on Clone, and change Into
+	// to AsRef, if possible
+	Identity: Into<DidIdentifierOf<T>> + Clone,
+{
+	fn decrement(&mut self) -> frame_support::dispatch::Weight {
+		Pallet::<T>::decrement_consumers_unsafe(&self.0 .1.clone().into());
+		T::WeightInfo::decrement_consumers()
+	}
+}
+
+impl<T: Config, Identity> IdentityConsumer<Identity, u32> for Pallet<T>
```

Comment:
I would call it Provider since the DidPallet provides identities and doesn't consume them

Reply:
The way Parity calls them is consumer for anything that depends on the identity, and provider for anything that make the identity usable. In this case I think we are counting the number of consumers so that the identity cannot be removed if > 0, not interested about whether the identity is usable or not (providers > 0).

Reply:
But I think you are right in a sense since this would be more like an `IdentityManager`, which could handle both providers and consumers. Will probably have to think of a better naming 😄